free/src/test/scala/cats/free/FreeApplicativeTests.scala

```diff
@@ -126,3 +129,44 @@ class FreeApplicativeTests extends CatsSuite {
     z.analyze(asString) should === ("xy")
   }
 }
+
+object FreeApplicativeTests {
+  private def freeGen[F[_], A](maxDepth: Int)(implicit F: Arbitrary[F[A]], FF: Arbitrary[(A, A) => A], A: Arbitrary[A]): Gen[FreeApplicative[F, A]] = {
+    val noFlatMapped = Gen.oneOf(
+      A.arbitrary.map(FreeApplicative.pure[F, A]),
+      F.arbitrary.map(FreeApplicative.lift[F, A]))
+
+    val nextDepth = Gen.chooseNum(1, math.max(1, maxDepth - 1))
+
+    def withFlatMapped = for {
+      fDepth <- nextDepth
+      freeDepth <- nextDepth
+      ff <- FF.arbitrary
+      f <- freeGen[F, A](fDepth).map(_.map(l => (u: A) => ff(l, u)))
+      freeFA <- freeGen[F, A](freeDepth)
+    } yield freeFA.ap(f)
+
+    if (maxDepth <= 1) noFlatMapped
+    else Gen.oneOf(noFlatMapped, withFlatMapped)
+  }
+
+  implicit def freeArbitrary[F[_], A](implicit F: Arbitrary[F[A]], FF: Arbitrary[(A, A) => A], A: Arbitrary[A]): Arbitrary[FreeApplicative[F, A]] =
+    Arbitrary(freeGen[F, A](2))
```

Comment:
Sorry I meant to ask this last time but somehow lost it. Did you try `maxDepth` larger than 2? The main reason I ask here is that it seemed to me that you were preparing for testing maxDepth more than 2.